modules/drivers/sqlite/src/metabase/driver/sqlite.clj

```diff
@@ -360,6 +361,97 @@
+                 (hsql/call
+                  :and
+                  (hsql/call := (extract :month-of-year a) (extract :month-of-year b))
+                  (hsql/call :> (extract :day-of-month a) (extract :day-of-month b)))))))]
+        (hsql/call :case (hsql/call :<= x y) (positive-diff x y) :else (hx/* -1 (positive-diff y x))))
+
+      :quarter
+      (let [positive-diff
+            (fn [a b]
+              (hx/cast
+               :integer
+               (hx/floor (hx// (hx/- (hx/+ (hx/* (hx/- (extract :year b)
+                                                       (extract :year a))
+                                                 12)
+                                           (hx/- (extract :month-of-year b)
+                                                 (extract :month-of-year a)))
+                                     (hx/cast
+                                      :integer
+                                      (hsql/call :> (extract :day-of-month a) (extract :day-of-month b))))
+                               3))))]
+        (hsql/call :case (hsql/call :<= x y) (positive-diff x y) :else (hx/* -1 (positive-diff y x))))
+
+      :month
+      (let [positive-diff
+            (fn [a b]
+              (hx/- (hx/+ (hx/* (hx/- (extract :year b)
+                                      (extract :year a))
+                                12)
+                          (hx/- (extract :month-of-year b)
+                                (extract :month-of-year a)))
+                    (hx/cast :integer (hsql/call :> (extract :day-of-month a) (extract :day-of-month b)))))]
+        (hsql/call :case (hsql/call :<= x y) (positive-diff x y) :else (hx/* -1 (positive-diff y x))))
+
+      :week
+      (let [positive-diff
+            (fn [a b]
+              (hx/cast :integer (hx// (hsql/call :-
+                                                 (hsql/call :julianday (->date b))
+                                                 (hsql/call :julianday (->date a)))
+                                      7)))]
+        (hsql/call :case (hsql/call :<= x y) (positive-diff x y) :else (hx/* -1 (positive-diff y x))))
+
+      :day
+      (let [positive-diff
+            (fn [a b]
+              (hx/cast :integer (hsql/call :-
+                                           (hsql/call :julianday (->date b))
+                                           (hsql/call :julianday (->date a)))))]
+        (hsql/call :case (hsql/call :<= x y) (positive-diff x y) :else (hx/* -1 (positive-diff y x))))
+
+      (:hour :minute :second)
+      (let [positive-diff
+            (fn [a b]
+              (-> (hsql/call :-
+                             (hsql/call :julianday b)
+                             (hsql/call :julianday a))
+                  (hx/* (case unit :hour 24 :minute 1440 :second 86400))
+                  hx/floor
+                  hx/->integer))]
+        (hsql/call :case (hsql/call :<= x y) (positive-diff x y) :else (hx/* -1 (positive-diff y x)))))))
```

Comment:
More than any of the other implementations I've looked at, this one would benefit from a `datetime-diff` multimethod. 90 line functions are way too long